botwo/botwo.py

```diff
@@ -0,0 +1,188 @@
+                client_to_call = res[0]
+                api_params = res[1]
+
+            return getattr(client_to_call, operation_name)(**api_params)
+
+        _api_call.__name__ = str(operation_name)
+        return _api_call
+
+    def _create_copy_method(self, operation_name):
+        def _api_call(*args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            client_to_call_source = self.default_client
+            client_to_call_dest = self.default_client
+            if kwargs.get("CopySource"):  # copy operation
+                res = self._route_params(kwargs["CopySource"], kwargs, is_copy=True)
+                client_to_call_source = res[0]
+                api_params = res[1]
+
+            if api_params.get("Bucket"):
+                full_dest_path = api_params["Bucket"] + "/" + api_params["Key"]
+                res = self._route_params(full_dest_path, api_params, is_copy=False)
+                client_to_call_dest = res[0]
+                api_params = res[1]
+
+            if client_to_call_source != client_to_call_dest:
+                raise ValueError("client source and client destination are different")
+
+            return getattr(client_to_call_source, operation_name)(**api_params)
+
+        _api_call.__name__ = str(operation_name)
+        return _api_call
+
+    def _create_special_method(self, operation_name):
+        def _paginator_api_call(*args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            return Paginatwo(self.mapping, kwargs, self.client1, self.client2, self.client1_name, self.client2_name)
+
+        def _delete_objects_api_call(*args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            # TODO
+            return None
+
+        if operation_name == "get_paginator":
+            _paginator_api_call.__name__ = str(operation_name)
+            return _paginator_api_call
+        if operation_name == "delete_objects":
+            _delete_objects_api_call.__name__ = str(operation_name)
+            return _delete_objects_api_call
+
+    def _route_params(self, path, api_params, is_copy):
+        client_to_call = self.default_client
+        for map in self.mapping:
+            if fnmatch.fnmatch(path, map.get("bucket_name")):
+                if map.get("client"):
+                    if map.get("client") == self.client2_name:
+                        client_to_call = self.client2
+                    elif map.get("client") == self.client1_name:
+                        client_to_call = self.client1
+                    else:
+                        raise ValueError("Invalid client name " + map.get("client"))
+                else:
+                    raise ValueError("client isn't provided")
```

Comment:
I think this logic is out of scope for _route_params?

Reply:
Done